File: gems/fact-mine/lib/fact_mine/syntax/protocols.rb
# frozen_string_literal: true

module FactMine
  module Syntax
    ProtocolMethodEffect = Struct.new(:file, :owner, :name, :line, :reads, :writes,
                                      keyword_init: true)
    ProtocolCall = Struct.new(:mid, :file, :owner, :defn, :line, :span, keyword_init: true)
    ProtocolMethodPath = Struct.new(:file, :owner, :name, :line, :calls, keyword_init: true)
    ProtocolPath = Struct.new(:calls, :terminal, keyword_init: true)
    ProtocolLexicon = Struct.new(
      :path_limit, :declarative_mids, :test_dsl_mids, :ignored_mids,
      :optional_diagnostic_mids, :mutating_mids, :non_mutating_operator_mids,
      :mutating_suffixes,
      keyword_init: true
    )

    EMPTY_PROTOCOL_LEXICON = ProtocolLexicon.new(
      path_limit: 64,
      declarative_mids: [].freeze,
      test_dsl_mids: [].freeze,
      ignored_mids: [].freeze,
      optional_diagnostic_mids: [].freeze,
      mutating_mids: [].freeze,
      non_mutating_operator_mids: [].freeze,
      mutating_suffixes: [].freeze
    ).freeze

    @protocol_lexicons = {}

    def self.register_protocol_lexicon(language, lexicon)
      @protocol_lexicons[language.to_sym] = lexicon
    end

    def self.protocol_lexicon_for(language)
      @protocol_lexicons.fetch(language.to_sym, EMPTY_PROTOCOL_LEXICON)
    end

    def self.protocol_ignored_mids(language)
      protocol_lexicon_for(language).ignored_mids
    end

    class Document
      def protocol_method_effects
        @protocol_method_effects ||= adapter.protocol_method_effects(self)
      end

      def protocol_call_paths
        @protocol_call_paths ||= adapter.protocol_call_paths(self)
      end
    end

    class TreeSitterLanguageAdapter
      def protocol_method_effects(document)
        behavior = Syntax::NormalizedExtractionBehavior.for(document.language)

        document.function_defs.map do |function_def|
          reads = document.state_reads.select do |read|
            read.owner == function_def.owner && read.function == function_def.name
          end.filter_map { |read| behavior.protocol_read_label_from_state(read) }

          reads.concat(document.call_sites.select do |call|
            call.owner == function_def.owner && call.function == function_def.name
          end.filter_map do |call|
            label = behavior.protocol_read_label_from_call(call)
            next nil if label.to_s.empty?

            label
          end)
          reads = reads.uniq.sort

          writes = document.state_writes.select do |write|
            write.owner == function_def.owner && write.function == function_def.name
          end.filter_map { |write| behavior.protocol_write_label(write) }.uniq.sort

          ProtocolMethodEffect.new(
            file: function_def.file,
            owner: function_def.owner,
            name: function_def.name.to_s.split(/[.:]/).last,
            line: function_def.line,
            reads: reads,
            writes: writes
          )
        end
      end

      def protocol_call_paths(document)
        document.function_defs.map do |function_def|
          calls = protocol_self_calls(document, function_def)

          protocol_call_variants(calls, function_def).map do |path_calls|
            ProtocolMethodPath.new(
              file: function_def.file,
              owner: function_def.owner,
              name: function_def.name.to_s.split(/[.:]/).last,
              line: function_def.line,
              calls: path_calls
            )
          end
        end.flatten
      end

      def protocol_self_calls(document, function_def)
        document.call_sites.select do |call|
            call.owner == function_def.owner &&
              call.function == function_def.name &&
              call.receiver.to_s == "self"
        end.map do |call|
          ProtocolCall.new(
            mid: call.message.to_s.split(/[.:]/).last,
            file: call.file,
            owner: call.owner,
            defn: call.function,
            line: call.line,
            span: call.span
          ).tap do |protocol_call|
            protocol_call.define_singleton_method(:conditional?) { call.conditional }
          end
        end
      end

      def protocol_call_variants(calls, function_def)
        return [[], []] if calls.empty? && protocol_conditional_body?(function_def.body)

        conditional_calls, always_calls = calls.partition do |call|
          call.respond_to?(:conditional?) && call.conditional?
        end
        return [calls] if conditional_calls.empty?

        conditional_calls.map { |call| (always_calls + [call]).sort_by(&:line) } +
          [always_calls.sort_by(&:line)]
      end

      def protocol_conditional_body?(node)
        return false unless node.is_a?(Hash)
        return true if %w[if unless case].include?(node["kind"].to_s)

        Array(node["children"]).any? { |child| protocol_conditional_body?(child) }
      end
    end

    class TreeSitterAdapter
      def protocol_method_effects(document)
        syntax_profile(document.language).protocol_method_effects(document)
      end

      def protocol_call_paths(document)
        syntax_profile(document.language).protocol_call_paths(document)
      end
    end
  end
end

File: gems/fact-mine/rust/src/syntax/protocols.rs
use super::{
    normalized_behavior::NormalizedLanguageBehavior, CallSite, FunctionDef, ProtocolCall,
    ProtocolMethodEffect, ProtocolMethodPath, SemanticEffectSite, StateRead, StateWrite,
};

pub(crate) fn method_effects_from_facts(
    behavior: &dyn NormalizedLanguageBehavior,
    function_defs: &[FunctionDef],
    state_reads: &[StateRead],
    state_writes: &[StateWrite],
    call_sites: &[CallSite],
    semantic_effect_sites: &[SemanticEffectSite],
) -> Vec<ProtocolMethodEffect> {
    function_defs
        .iter()
        .map(|function_def| {
            let mut reads = state_reads
                .iter()
                .filter(|read| {
                    read.owner == function_def.owner && read.function == function_def.name
                })
                .filter_map(|read| {
                    behavior.protocol_read_label_from_state(&read.receiver, &read.field)
                })
                .collect::<Vec<_>>();
            reads.extend(
                call_sites
                    .iter()
                    .filter(|call| {
                        call.owner == function_def.owner
                            && call.function == function_def.name
                            && !semantic_effect_call(semantic_effect_sites, call)
                    })
                    .filter_map(|call| {
                        behavior.protocol_read_label_from_call(&call.receiver, &call.message)
                    }),
            );
            reads.sort();
            reads.dedup();

            let mut writes = state_writes
                .iter()
                .filter(|write| {
                    write.owner == function_def.owner && write.function == function_def.name
                })
                .filter_map(|write| behavior.protocol_write_label(&write.receiver, &write.field))
                .collect::<Vec<_>>();
            writes.sort();
            writes.dedup();

            ProtocolMethodEffect {
                file: function_def.file.clone(),
                owner: function_def.owner.clone(),
                name: protocol_method_name(&function_def.name),
                line: function_def.line,
                reads,
                writes,
            }
        })
        .collect()
}

pub(crate) fn call_paths_from_facts(
    behavior: &dyn NormalizedLanguageBehavior,
    function_defs: &[FunctionDef],
    call_sites: &[CallSite],
    semantic_effect_sites: &[SemanticEffectSite],
) -> Vec<ProtocolMethodPath> {
    let mut out = Vec::new();
    for function_def in function_defs {
        let calls = protocol_self_calls(behavior, function_def, call_sites, semantic_effect_sites);
        let variants = if calls.is_empty() {
            protocol_empty_variants_for_function(function_def).unwrap_or_else(|| vec![Vec::new()])
        } else {
            protocol_call_variants(calls)
        };
        for calls in variants {
            out.push(ProtocolMethodPath {
                file: function_def.file.clone(),
                owner: function_def.owner.clone(),
                name: protocol_method_name(&function_def.name),
                line: function_def.line,
                calls,
            });
        }
    }
    out
}

fn protocol_empty_variants_for_function(
    function_def: &FunctionDef,
) -> Option<Vec<Vec<ProtocolCall>>> {
    if protocol_conditional_body(&function_def.body) {
        Some(vec![Vec::new(), Vec::new()])
    } else {
        None
    }
}

fn protocol_call_variants(calls: Vec<(ProtocolCall, bool)>) -> Vec<Vec<ProtocolCall>> {
    let (conditional_calls, always_calls): (Vec<_>, Vec<_>) =
        calls.into_iter().partition(|(_, conditional)| *conditional);
    let always_calls = always_calls
        .into_iter()
        .map(|(call, _)| call)
        .collect::<Vec<_>>();
    if conditional_calls.is_empty() {
        return vec![always_calls];
    }

    let mut out = conditional_calls
        .into_iter()
        .map(|(conditional_call, _)| {
            let mut path = always_calls.clone();
            path.push(conditional_call);
            path.sort_by_key(|call| call.line);
            path
        })
        .collect::<Vec<_>>();
    let mut always_path = always_calls;
    always_path.sort_by_key(|call| call.line);
    out.push(always_path);
    out
}

fn protocol_method_name(name: &str) -> String {
    name.split(['.', ':'])
        .next_back()
        .unwrap_or(name)
        .trim_end_matches(['?', '!'])
        .to_string()
}

fn protocol_conditional_body(node: &crate::ast::RawNode) -> bool {
    let kind = node.kind.to_ascii_uppercase();
    matches!(kind.as_str(), "IF" | "UNLESS" | "CASE" | "CASE2" | "WHEN")
        || node.children.iter().any(protocol_conditional_body)
}

fn protocol_self_calls(
    behavior: &dyn NormalizedLanguageBehavior,
    function_def: &FunctionDef,
    call_sites: &[CallSite],
    semantic_effect_sites: &[SemanticEffectSite],
) -> Vec<(ProtocolCall, bool)> {
    call_sites
        .iter()
        .filter(|call| {
            call.owner == function_def.owner
                && call.function == function_def.name
                && !semantic_effect_call(semantic_effect_sites, call)
        })
        .filter_map(|call| {
            protocol_call(behavior, function_def, call).map(|row| (row, call.conditional))
        })
        .collect()
}

fn protocol_call(
    behavior: &dyn NormalizedLanguageBehavior,
    function_def: &FunctionDef,
    call: &CallSite,
) -> Option<ProtocolCall> {
    let mid = behavior.protocol_read_label_from_call(&call.receiver, &call.message)?;
    Some(ProtocolCall {
        mid: protocol_method_name(&mid),
        file: function_def.file.clone(),
        owner: function_def.owner.clone(),
        defn: protocol_method_name(&function_def.name),
        line: call.line,
        span: call.span,
    })
}

fn semantic_effect_call(semantic_effect_sites: &[SemanticEffectSite], call: &CallSite) -> bool {
    semantic_effect_sites.iter().any(|effect| {
        effect.function == call.function && effect.line == call.line && effect.span == call.span
    })
}
